GowerMetric.py

```diff
@@ -0,0 +1,371 @@
+
+    def __call__(
+        self,
+        vector_1: np.ndarray,
+        vector_2: np.ndarray,
+    ):
+        # TODO - no check for fit
+        # sklearn.utils.validation.check_is_fitted(self)
+        if self.n_features_in_ != len(vector_1) | len(vector_2):
+            assert len(vector_1) == len(vector_2)
+
+        # Distance for cat_nom and bin_sym columns
+        if self._cat_nom_num:
+            cat_nom_cols_1 = vector_1[self._cat_nom_idx]
+            cat_nom_cols_2 = vector_2[self._cat_nom_idx]
+            cat_nom_dist = 1.0 - (cat_nom_cols_1 == cat_nom_cols_2)
+
+            # Weighted variant - dot product with according self.weights columns
+            # Unweighted - a sum of distances of between all columns
+            if self.weights is not None:
+                cat_nom_dist = cat_nom_dist @ self.weights[self._cat_nom_idx]
+            else:
+                cat_nom_dist = cat_nom_dist.sum()
+        else:
+            cat_nom_dist = 0.0
+
+        # Distance for ratio_scale columns
+        if self._ratio_scale_num:
+            ratio_scale_cols_1 = vector_1[self._ratio_scale_idx]
+            ratio_scale_cols_2 = vector_2[self._ratio_scale_idx]
+            ratio_dist = np.abs(ratio_scale_cols_1 - ratio_scale_cols_2)
+
+            ones = ratio_dist >= self.ranges_[self._ratio_scale_idx]
+            zeros = ratio_dist <= self.h_[self._ratio_scale_idx]
+
+            if ratio_scale_cols_1.size == 1:
+                if ones:
+                    ratio_dist = 1.0
+                elif zeros:
+                    ratio_dist = 0.0
+                else:
+                    ratio_dist /= self.ranges_[self._ratio_scale_idx]
+            else:
+                ratio_dist /= self.ranges_[self._ratio_scale_idx]
+                ratio_dist[zeros] = 0.0
+                ratio_dist[ones] = 1.0
+
+            if self.weights is not None:
+                ratio_dist = ratio_dist @ self.weights[self._ratio_scale_idx]
+            else:
+                ratio_dist = ratio_dist.sum()
+        else:
+            ratio_dist = 0.0
+
+        # Sum of all distances
+        distance = cat_nom_dist + ratio_dist
+
+        # Normalization
+        distance /= self.n_features_in_
+
+        return distance
+
+    def fit(self, X):
+        self.ranges_ = np.zeros(self.dtypes.size)
+        self.h_ = np.zeros(self.dtypes.size)
+        self.n_features_in_ = self.dtypes.size
+
+        for i in range(self.dtypes.size):
+            if self.dtypes[i] == DataType.RATIO_SCALE:
+                column = X[:, i]
+
+                if self.ratio_scale_normalization == "iqr":
+                    # IQR (g_t) - Interquartile Range
+                    q1, q3 = np.percentile(column, [25, 75])
+                    self.ranges_[i] = q3 - q1
+
+                    if self.ranges_[i] == 0:
+                        self.ranges_[i] = np.ptp(column, axis=0)
+
+                elif self.ratio_scale_normalization == "range":
```

Comment:
This should be before "iqr", since it's more typical. In general options in if/else should be ordered from the most common to the least common for both readability and efficiency.